csrc/id_graphs.cpp

```diff
@@ -0,0 +1,3453 @@
+
+  // Promotion logic is going to be on the intersection of the exact and loop
+  // graph. We will generate a map on the entries of this graph so it's
+  // important to not modify this graph moving forward, as that would invalidate
+  // the map.
+  //
+  // iel stands for Intersection of the Exact and Loop graphs.
+  std::unordered_map<IdGroup, IterDomain*> iel_promotion_map;
+
+  // This should probably work just on terminating inputs, as we shouldn't be
+  // able to modify a broadcast domain between root and rfactor which would be
+  // required to resolve a non input broadcast domain. But for now leaving it as
+  // traversal on all broadcast groups.
+  for (auto iel_group :
+       intersection_exact_loop_graph.disjointIdSets().disjointSets()) {
+    if (!iel_group->front()->isBroadcast()) {
+      continue;
+    }
+
+    // Collect all the exact groups of the resolutions of the broadcast id's
+    IdGroups resolved_exact_groups;
+    for (auto bcast_id : *iel_group) {
+      auto p2c_root_broadcast_resolution_map_it =
+          p2c_root_broadcast_resolution_map.find(bcast_id);
+
+      if (p2c_root_broadcast_resolution_map_it ==
+          p2c_root_broadcast_resolution_map.end()) {
+        continue;
+      }
+
+      resolved_exact_groups.pushBack(
+          idGraph(IdMappingMode::EXACT)
+              .toGroups(p2c_root_broadcast_resolution_map_it->second));
+    }
+
+    // Collect all the exact groups in the loop set containing this iel_group
+    auto loop_group_pair =
+        idGraph(IdMappingMode::LOOP).disjointIdSet(iel_group->front());
+    TORCH_INTERNAL_ASSERT(loop_group_pair.second);
+    auto loop_group = loop_group_pair.first;
+    auto loop_covered_exact_groups =
+        idGraph(IdMappingMode::EXACT).toGroups(*loop_group);
+
+    // The intersection of the exact groups that the broadcast domains can be
+    // broadcasted to, and those that exist within the same loop are is the
+    // promotion needed for this iel_group.
+    auto loop_exact_resolved_intersection =
+        resolved_exact_groups.intersect(loop_covered_exact_groups);
+
+    if (loop_exact_resolved_intersection.empty()) {
+      // No resolution
+      continue;
+    }
+
+    if (loop_exact_resolved_intersection.size() > 1) {
+      std::stringstream err_msg;
+
+      err_msg
+          << "Invalid multiple broadcast resolution within shared loops detected, group:\n  "
+          << iel_group->toString() << "\nIs being broadcasted to:";
+
+      for (auto entry : loop_exact_resolved_intersection) {
+        err_msg << "\n  " << entry->toString();
+      }
+      TORCH_INTERNAL_ASSERT(false, err_msg.str());
+    }
+
+    // loop_exact_resolved_intersection.size() == 1
+    auto exact_resolution_group = loop_exact_resolved_intersection.front();
+
+    VectorOfUniqueEntries<IterDomain*> resolved_ids =
+        exact_resolution_group->intersect(*loop_group);
+    auto promoted_iel_groups =
+        intersection_exact_loop_graph.toGroups(resolved_ids);
```

Comment:
Is this necessary? `resolved_ids` is an intersection of `exact_resolution_group` and `loop_group`, so by definition it should be mapped by both the exact and loop maps, right?

Reply:
`loop_exact_resolved_intersection` is in the exact id graph, we want the promotion group to be in the iel graph, that's where the replay will occur.

`exact_resolution_group->intersect(*loop_group)` will remove any exact id that's not within this loop. So if we then call `intersection_exact_loop_graph.toGroups(resolved_ids)` having more than one group would be a logical inconsistency error. So we should be able to do:
`intersection_exact_loop_graph.toGroups({resolved_ids.front()}}).front()` or whatever the right syntax is.

Reply:
One way or another we do need the intersection, and that intersection needs to be described in the iel graph.

Reply:
Thanks. I realized I had a misunderstanding.